### DISCUSSION TITLE
Booting A/B, where is the kernel used for booting device?

### DISCUSSION THREAD

**Discussion by AndersGnistrup** [OE4T/meta-tegra#1594] [General]
I have a working system so this is just a question about "how stuff works".
I am using branch scarthgap-l4t-r35.x on tegra-demo-distro and have found similar result for other branches. The development board is a orin-nano 8Gb NVidia development board. That branch was selected to get back to a known state.

I have inspecting the partition layout and from the file flash_l4t_t234_nvme_rootfs_ab.xml there should be a redundant setup. That also seems to match the logs during flashing
```
== Step 4: Writing partitions on external storage device at 2024-06-24T07:04:57+02:00 ==
Waiting for USB storage device nvme0n1 from a92bfe0a...[/dev/sda]
Creating partitions
  [03] name=A_kernel start=0 size=262144 sectors
  [04] name=A_kernel-dtb start=0 size=1536 sectors
  [05] name=A_reserved_on_user start=0 size=64768 sectors
  [06] name=B_kernel start=0 size=262144 sectors
  [07] name=B_kernel-dtb start=0 size=1536 sectors
  [08] name=B_reserved_on_user start=0 size=64768 sectors
  [09] name=recovery start=0 size=163840 sectors
  [10] name=recovery-dtb start=0 size=1024 sectors
  [11] name=esp start=0 size=131072 sectors
  [12] name=recovery_alt start=0 size=163840 sectors
  [13] name=recovery-dtb_alt start=0 size=1024 sectors
  [14] name=esp_alt start=0 size=131072 sectors
  [15] name=UDA start=0 size=819200 sectors
  [16] name=reserved start=0 size=982016 sectors
  [01] name=APP start=0 size=29360128 sectors
  [02] name=APP_b start=0 size=29360128 sectors

```
The APP has id=1 and APP_b has a id=2 so it seems to match the xml partition layout description in the xml file.

So far, so good, but I had a hard time to verify the boot selection/boot order using the boot log and system inspection.
So I decided to clear the A_kernel partition on a running system using `dd if=/dev/zero of=/dev/nvme0n1p3`.
I expected a boot failure/switch to B_kernel since the kernel was gone. But, the system booted without any delay.
Then I cleared the B_kernel, to see if some magic fallback had occurred. But, the system still booted without any delay.
Finally, I tried to clear the APP and that "worked". After three boot attempts in UEFI it switched to APP_b.

```
root@jetson-orin-nano-devkit-nvme:~# df
Filesystem           1K-blocks      Used Available Use% Mounted on
none                   3681192         0   3681192   0% /dev
/dev/nvme0n1p2        13655048    776516  12128148   6% /
tmpfs                  3815308         0   3815308   0% /dev/shm
tmpfs                  1526124      9668   1516456   1% /run
tmpfs                  3815308         0   3815308   0% /tmp
tmpfs                  3815308         8   3815300   0% /var/volatile
/dev/nvme0n1p11          64511       130     64382   0% /boot/efi
/dev/nvme0n1p11          64511       130     64382   0% /opt/nvidia/esp
tmpfs                   763060         4    763056   0% /run/user/0

```

But, the system should not boot at all since there is no kernel ?
Is something "not quite ready" but layout is kept for future use.
The running kernel is not in in sda3 but "somewhere else", probably in the rootfs partition?

**Comment by AndersGnistrup** [comment]
Moved to tegre-demo-distro.
Sorry.

**Reply by dwalkes** [reply]
Cross linking, thread at https://github.com/OE4T/tegra-demo-distro/discussions/301
